h_malloc.c

```diff
@@ -1619,11 +1740,13 @@ static inline void memory_corruption_check_small(const void *p) {
     mutex_unlock(&c->lock);
 }
 
-EXPORT size_t h_malloc_usable_size(H_MALLOC_USABLE_SIZE_CONST void *p) {
-    if (p == NULL) {
+EXPORT size_t h_malloc_usable_size(H_MALLOC_USABLE_SIZE_CONST void *arg) {
+    if (arg == NULL) {
         return 0;
     }
 
+    void *p = untag_pointer((void *) (uintptr_t) arg);
```

Comment:
`arg` is already a `void*`, why cast it to an `uintptr_t` then back to `void*`?

Reply:
arg is `const void *` on Android. Haven't found a better solution to prevent const-correctness checks from breaking the build.

Reply:
Isn't this [undefined behaviour](https://wiki.sei.cmu.edu/confluence/display/c/EXP05-C.+Do+not+cast+away+a+const+qualification )?

Reply:
Will try building this on the LLVM RC to see what it says about that, but these sorts of hack-arounds seem to end up making stricter compiler checks unhappy over time (try building NV's driver code even with LLVM's kCFI - sadness occurs). Would it be more correct to use a separate non-constified variable taking the const'd `arg`s value?

Reply:
Casting away `const` and modifying it isn't undefined. It's only undefined if the underlying data can't be modified. `const` pointers (as in the data is read-only through the pointer) are only a programming aid with no actual functionality.

Reply:
> Would it be more correct to use a separate non-constified variable taking the const'd args value?

This is detected by the compiler too IIRC

Reply:
Then a comment should be added, because it's a tad confusing :)

Reply:
Moved the cast into the new untag_const_pointer() function: https://github.com/GrapheneOS/hardened_malloc/pull/225/commits/452ff8a878826281bdfee81a1463b33ff4e889e0